jax/experimental/jax2iree/primitives.py

```diff
@@ -0,0 +1,98 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from jax import core
+from typing import Any, Callable, Dict, Sequence, Union
+
+from jax._src.lax import lax
+
+from .ir_builder import FunctionBuilder
+from .mlir_imports import *
+
+
+class PrimitiveInvocation:
+  __slots__ = [
+      "primitive",
+      "vals",
+      "avals",
+      "out_aval",
+      "params",
+  ]
+
+  def __init__(self, primitive: core.Primitive, vals: Sequence[ir.Value],
+               avals: Sequence[core.AbstractValue],
+               out_aval: Union[core.AbstractValue,
+                               Sequence[core.AbstractValue]], params: dict):
+    self.primitive = primitive
+    self.vals = vals
+    self.avals = avals
+    self.out_aval = out_aval
+    self.params = params
+
+  def emit_fallback(self, fb: FunctionBuilder):
+    # TODO: This is super fragile: just to see what is going on.
+    op_name = f"jax_unrecognized.{self.primitive.name}"
```

Comment:
What is this used for? Oh yes I think you mentioned this is just for testing to avoid triggering any optimizations etc while bootstrapping. Is that correct?

Reply:
Done right, it would be a real debugging aid: For unrepresented ops, actually emit them for inspection (with location info) vs hard-erroring on the first such and making someone parse python exception logs.